python/ray/tests/test_basic.py

```diff
@@ -2843,3 +2843,10 @@ def test_load_code_from_local(shutdown_only):
     # Test derived actor class.
     actor = DerivedClass.remote(1)
     assert ray.get(actor.get_data.remote()) == 1
+
+
+def test_connect_with_disconnected_node(ray_start_cluster):
```

Comment:
I think this should go in `python/ray/tests/test_multi_node_2.py`. Did you add the `allow_graceful` flag for consistency or for some other reason? It doesn't seem to be used anywhere.

Reply:
Yes, this flag is only used for this test. This test is used to repro the problem @hartikainen mentioned. I can check the raylet log to see whether there is `Failed to connect to client`. However, this required `allow_graceful=True` to send `SIGTERM` to raylet instead of `SIGKILL`. Actually, even without `allow_graceful=True` the test also passes due to https://github.com/ray-project/ray/pull/2905 . This option is used for reproducing. I can remove the flag after this problem fixed.
Do you have any comment for adding a callback to close the main service to guarantee the closing entry is added to GCS?

Reply:
@robertnishihara I think the option `allow_graceful` of `ray.tests.cluster_utils` will be kept to make the test case work.